keps/sig-node/5067-pod-generation/README.md

```diff
@@ -0,0 +1,930 @@
+* Expand the set of mutable fields.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### API Changes
+
+#### Generation and ObservedGeneration
+
+The `metadata.generation` field is currently unused on pods, but we can start 
+leveraging it to help track which pod state is currently being reflected in the 
+pod status. [Generation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata) is documented as "a sequence 
+number representing a specific generation of the desired state. Set by the 
+system and monotonically increasing, per-resource." For consistency, pod 
+Generation will be incremented whenever the pod has changes that the kubelet 
+needs to actuate. 
+
+A new field ObservedGeneration will be added to the pod status. 
+Kubelet will set this to communicate which pod state is being expressed in the 
+current pod status. This is analogous to the ObservedGeneration that exists
+in other resources' statuses such as [StatefulSets](https://github.com/kubernetes/api/blob/9e7d345b161c12a3056efa88e285b3ef68450c54/apps/v1/types.go#L274).
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+No risks have been identified.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API server
+
+For a newly created pod, the API server will set `generation` to 1. For every 
+update to the pod that requires actuation by kubelet, the API Server will 
+increment `generation` by 1. Pod updates that result in `generation` being 
+incremented are any changes to mutable fields in the [PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core). 
+
+Pod updates that would not result in `generation` being incremented include:
```

Comment:
If a Pod uses the downward API to make Pod metadata available to containers, Pod behavior may change without the generation changing.


(OK, so long as we expect that; might need special care to document it).

Reply:
(optionally, file this as an actual alternative)

We can detect downward API references to fields and bump generation only if a referenced field changes.